tests/test_file_contents.py

```diff
@@ -0,0 +1,157 @@
+import pathlib
+import time
+from collections import deque
+from unittest.mock import MagicMock, Mock
+
+import pytest
+import watchfiles
+from watchfiles import Change
+
+from anywidget._file_contents import FileContents
+
+
+def test_file_contents_no_watch(tmp_path: pathlib.Path):
+    """Test __str__ reads file contents and does not start a thread"""
+    CONTENTS = "hello, world"
+    path = tmp_path / "foo.txt"
+    path.write_text(CONTENTS)
+
+    with patch.object(watchfiles, "watch") as mock:
+        contents = FileContents(path, start_thread=False)
+        assert str(contents) == CONTENTS
+        assert contents._background_thread is None
+    mock.assert_not_called()
+
+
+def test_file_contents_deleted(monkeypatch: pytest.MonkeyPatch, tmp_path: pathlib.Path):
+    """Test deleting a file emits a deleted signal and stops the watcher"""
+    CONTENTS = "hello, world"
+    path = tmp_path / "foo.txt"
+    path.write_text(CONTENTS)
+
+    def mock_file_events():
+        changes = set()
+        changes.add((Change.deleted, str(path)))
+        yield changes
+        changes = set()
+        changes.add((Change.modified, str(path)))
+        yield changes
+
+    mock_watch = MagicMock(return_value=mock_file_events())
+    monkeypatch.setattr(watchfiles, "watch", mock_watch)
+
+    contents = FileContents(path, start_thread=False)
+    mock = Mock()
+    contents.deleted.connect(mock)
+
+    total = sum(1 for _ in contents.watch())
+    assert total == 0
+    mock_watch.assert_called_with(path, stop_event=contents._stop_event)
+    assert mock.called
```

Comment:
another optional alternative here:

```suggestion
def test_file_contents_deleted(tmp_path: pathlib.Path):
    """Test deleting a file emits a deleted signal and stops the watcher"""
    CONTENTS = "hello, world"
    path = tmp_path / "foo.txt"
    path.write_text(CONTENTS)

    contents = FileContents(path)

    mock = Mock()
    contents.deleted.connect(mock)
    path.unlink()
    assert contents._background_thread
    contents._background_thread.join()
    assert not contents._background_thread.is_alive()
    mock.assert_called()
```

this just lets watcher.watch do it's thing with an actual delete event (no mock)

Reply:
oh right, this is great!

Reply:
as a side note here... a handy thing in `unittest`, if you ever _want_ to let an object do its regular job, but also assert that it was indeed called, you can use the `wraps` argument.  so it would be something like:


```python
    with patch.object(watchfiles, "watch", wraps=watchfiles.watch) as mock_watch:
        contents = FileContents(path)
        mock_watch.assert_called_once_with(contents._path, stop_event=contents._stop_event)
```

Reply:
good to know. hmm this test is failing now. will poke around a bit later. I think this is related to an issue I was running into earlier where it took a little while to "start" watching so the unlink wasn't observed.

Reply:
hmm, passing locally... lemme check 3.8

Reply:
arg... you could try this:

```python
    contents = FileContents(path)
    while not contents._background_thread.is_alive():
        time.sleep(0.1)
```


or just go back to before :/